# Cost and Profitability Analysis:

Cost and Profitability analysis examines the costs and revenue involved in delivering food orders. 

It looks at expenses like delivery fees and discounts against income from orders and commissions to understand the service's profitability per order. Through careful data analysis and strategic recommendations, the project aims to improve efficiency and increase profits in food delivery operations.

#### Below is the process for conducting the Food Delivery Cost and Profitability Analysis:
1. Data Cleaning: Ensure data accuracy by removing inconsistencies and standardizing date/time formats in the dataset.
2. Feature Extraction: Identify key factors impacting cost and profitability, such as Payment Method, Delivery Fee, Discounts, etc.
3. Transaction Analysis: Analyze customer behavior and transactional patterns to discern trends and preferences.
4. Cost and Profitability Analysis: Compute profit and revenue from orders, considering order values and commissions. Visualize cost, revenue, and profit distributions to compare total revenue, costs, and profit.
5. Profit Enhancement Strategy: Determine an optimal balance between offering discounts and charging commissions. Develop new average commission and discount percentages for profitable orders.

In [1]:
# importing libraries:
import pandas as pd
from datetime import datetime
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.io as pio
import plotly.graph_objects as go
from plotly.subplots import make_subplots
pio.templates.default = "plotly_white"

In [2]:
data = pd.read_csv(r"C:\Users\user\Desktop\April\Cost and Profitability Analysis\Food Delivery Services.csv")
data.head(3)

Order ID Customer ID Restaurant ID  Order Date and Time  \
0         1       C8270         R2924  2024-02-01 01:11:52   
1         2       C1860         R2054  2024-02-02 22:11:04   
2         3       C6390         R2870  2024-01-31 05:54:35   

  Delivery Date and Time  Order Value  Delivery Fee    Payment Method  \
0    2024-02-01 02:39:52         1914             0       Credit Card   
1    2024-02-02 22:46:04          986            40    Digital Wallet   
2    2024-01-31 06:52:35          937            30  Cash on Delivery   

  Discounts and Offers  Commission Fee  Payment Processing Fee  \
0            5% on App             150                      47   
1                  10%             198                      23   
2         15% New User             195                      45   

   Refunds/Chargebacks  
0                    0  
1                    0  
2                    0

### The dataset comprises 1000 entries and encompasses various attributes related to food orders in a delivery service. These attributes include:

* Order ID: A unique identifier for each order.
* Customer ID: Identification of the customer placing the order.
* Restaurant ID: Identification of the restaurant fulfilling the order.
* Order Date and Time: Timestamp indicating when the order was placed.
* Delivery Date and Time: Timestamp indicating when the order was delivered.
* Order Value: The total value of the order.
* Delivery Fee: The fee charged for delivering the order.
* Payment Method: The method used by the customer to make payment.
* Discounts and Offers: Any discounts or promotional offers applied to the order.
* Commission Fee: The fee charged to the restaurant by the delivery service.
* Payment Processing Fee: The fee incurred for processing the payment.
* Refunds/Chargebacks: Any refunds or chargebacks associated with the order.

These attributes provide comprehensive information for analyzing the cost and profitability of the food delivery service, as well as understanding customer behavior and transactional patterns.

# Data Cleaning and Preparation:

In [3]:
data.isnull().sum()

Order ID                    0
Customer ID                 0
Restaurant ID               0
Order Date and Time         0
Delivery Date and Time      0
Order Value                 0
Delivery Fee                0
Payment Method              0
Discounts and Offers      185
Commission Fee              0
Payment Processing Fee      0
Refunds/Chargebacks         0
dtype: int64

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Order ID                1000 non-null   int64 
 1   Customer ID             1000 non-null   object
 2   Restaurant ID           1000 non-null   object
 3   Order Date and Time     1000 non-null   object
 4   Delivery Date and Time  1000 non-null   object
 5   Order Value             1000 non-null   int64 
 6   Delivery Fee            1000 non-null   int64 
 7   Payment Method          1000 non-null   object
 8   Discounts and Offers    815 non-null    object
 9   Commission Fee          1000 non-null   int64 
 10  Payment Processing Fee  1000 non-null   int64 
 11  Refunds/Chargebacks     1000 non-null   int64 
dtypes: int64(6), object(6)
memory usage: 93.9+ KB


* #### Convert “Order Date and Time” and “Delivery Date and Time” to a datetime format.

In [5]:
# convert date and time columns to datetime
data['Order Date and Time'] = pd.to_datetime(data['Order Date and Time'])
data['Delivery Date and Time'] = pd.to_datetime(data['Delivery Date and Time'])

* #### Convert “Discounts and Offers” to a consistent numeric value (percentage) and calculate the discount amounts.

In [6]:
# Convert the 'Discounts and Offers' column to strings
data['Discounts and Offers'] = data['Discounts and Offers'].astype(str)

# Apply the function to create a new 'Discount Percentage' column
data['Discount Percentage'] = data['Discounts and Offers'].apply(calculate_discount)

# Calculate the 'Discount Amount' based on the order value and discount percentage
data['Discount Amount'] = data.apply(lambda row: row['Order Value'] * row['Discount Percentage'], axis=1)

# Display the DataFrame
data[['Order Value', 'Discounts and Offers', 'Discount Percentage', 'Discount Amount']].head(5)


NameError: name 'calculate_discount' is not defined

In [ ]:
data.info()

In [ ]:
data.nunique()

# Order Transaction Analysis:

#### This analysis delves into various aspects of order transactions to gain insights into customer behavior and transactional patterns. 

#### Here's a breakdown of the key attributes considered:

* Delivery Fee: Understanding how delivery fees impact costs and customer behavior, including their willingness to pay for delivery.

* Payment Method: Exploring customer preferences and the popularity of different payment options to inform decisions on payment processing and optimization.

* Discounts and Offers: Assessing the effectiveness of promotional strategies and customer response to incentives, aiding in the evaluation of their impact on order value and profitability.

* Refunds/Chargebacks: Identifying patterns in refunds and chargebacks to pinpoint order issues, evaluate customer satisfaction, and gauge the effectiveness of dispute resolution processes.




By analyzing these attributes collectively, businesses can gain a comprehensive understanding of their order transactions, identify areas for improvement, and make data-driven decisions to enhance customer experience, optimize operations, and drive profitability.

In [ ]:
categories = ['Payment Method', 'Delivery Fee', 'Discount Percentage', 'Refunds/Chargebacks']
fig = make_subplots(
    rows=2,
    cols=2,
    specs=[[{"type": "domain"} for _ in range(2)] for _ in range(2)],
    #subplot_titles=categories,  # Set subplot titles
    vertical_spacing=0.2,  # Adjust vertical spacing between subplots
    horizontal_spacing=0.1,  # Adjust horizontal spacing between subplots
    shared_yaxes=True) # Share y-axes across subplots

# Loop through each category
for i, category in enumerate(categories, start=1):
    # Create pivot table
    pivot_table = pd.pivot_table(data, values='Order ID', index=[category], aggfunc='count')
    
    # Calculate total count for percentage calculation
    total_count = pivot_table['Order ID'].sum()
    
    # Iterate over pivot table rows and add pie chart trace to subplot
    labels = []
    for index, row in pivot_table.iterrows():
        percentage = row['Order ID'] / total_count * 100
        label = f"({index})"
        labels.append(label)
    
    fig.add_trace(go.Pie(values=pivot_table['Order ID'], labels=labels,textinfo='label+percent',
                         name=category, title=f'{category}',showlegend=False), row=(i-1)//2 + 1, col=(i-1)%2 + 1)

# Update layout
fig.update_layout(
    title='<b>Order Transaction Analysis</b>',
    title_x=0.45,
    height=900,  
    width=900)  

# Show the figure
fig.show()

#### Payment Method:
* Cash on Delivery leads as the most preferred payment method with 357 orders, followed closely by Credit Card (337 orders) and Digital Wallet (306 orders). 
* This underscores the importance of offering diverse payment options to accommodate varied customer preferences.

#### Delivery Fee:
* Customers seem to accept varying delivery fees as orders are evenly spread across different fee brackets (0, 20, 30, 40, and 50 INR). 
* This suggests that adjusting delivery fee structures strategically could boost revenue without much impact on order volumes

#### Discount Percentage:
* Moderate discounts of 10.0% (233 orders) and no discounts (185 orders) are common, but deeper discounts of 50.0% (201 orders) also appeal to a substantial portion of customers, which need to reduce for profitability. 
* Tailoring discount strategies based on these preferences can help achieve promotional goals effectively.

#### Refunds/Chargebacks:
* The majority of orders (715) experience no refunds or chargebacks, signaling effective dispute resolution and high customer satisfaction levels. 
* However, a small number of orders involve refunds/chargebacks of 50, 100, and 150 INR, underscoring the need for ongoing monitoring and enhancement of customer service and order fulfillment procedures.



# Cost and Profitability Analysis:

#### For analyzing costs, we'll look at what it takes to handle each order:

* Delivery Fee: The cost for delivering orders.
* Payment Processing Fee: The charge for processing payments.
* Discount Amount: Any discounts given on orders.

We'll add up all these costs for each order to understand how much it costs the platform overall (Total Cost).

The platform makes money mainly from commission fees (Revenue).

To find out how much profit it's making, we'll subtract all costs (including discounts) from the commission fees earned (Profit).

Now, let's dive into analyzing costs and profitability.

In [ ]:
# calculate total costs, revenue per order, profit:
data['Total Costs'] = data['Delivery Fee'] + data['Payment Processing Fee'] + data['Discount Amount']
data['Revenue'] = data['Commission Fee']
data['Profit'] = data['Revenue'] - data['Total Costs']

In [ ]:
# Calculate overall metrics:
total_orders = data.shape[0]
total_revenue = data['Revenue'].sum()
total_costs = data['Total Costs'].sum()
total_profit = data['Profit'].sum()

overall_metrics = {
    "Total Orders": total_orders,
    "Total Revenue": total_revenue,
    "Total Costs": total_costs,
    "Total Profit": total_profit
}
overall_metrics

### Here are the key findings from the analysis of the food delivery operations:

* Total Orders: 1,000
* Total Revenue (from Commission Fees): 126,990 INR
* Total Costs: 232,709.85 INR (includes delivery fees, payment processing fees, and discounts)
* Total Profit: -105,719.85 INR

The analysis reveals that the total costs exceed the revenue generated from commission fees, resulting in a net loss. 

This suggests that the current commission rates, delivery fees, and discount strategies may not be sustainable for profitability.

### To visualize the distribution of costs, revenue, and profit, we'll create the following plots:

* Histogram: Visualize the distribution of profits per order to see the mix of profitable and unprofitable orders.
* Pie Chart: Show the proportion of total costs (including delivery fees, payment processing fees, and discounts).
* Bar Chart: Compare total revenue, total costs, and total profit.

In [ ]:
# Create histogram
fig = px.histogram(data, x='Profit', histnorm='probability density', marginal='rug',
                   labels={'Profit': 'Profit', 'count': 'Number of Orders'},opacity=0.5,
                   title='<b>Profit Distribution per Order</b>')


# Add vertical line for mean profit
fig.add_vline(x=data['Profit'].mean(), line_dash="dash", line_color="red", 
              annotation_text="Mean Profit", annotation_position="top left")

fig.update_layout(
    title_x=0.5,  # Center the title horizontally
    title_y=0.95  # Position the title closer to the top of the figure
)

# Show the plot
fig.show()


* The histogram reveals a broad distribution of profit per order, with a significant portion of orders resulting in a loss (profits below 0).
* The red dashed line represents the average profit, which falls into negative territory, emphasizing the overall loss-making scenario.

#### Next, let's examine the proportion of total costs:

In [13]:
# Calculate total costs breakdown
costs_breakdown = data[['Delivery Fee', 'Payment Processing Fee', 'Discount Amount']].sum()

# Create pie chart
fig = go.Figure(data=[go.Pie(labels=costs_breakdown.index, values=costs_breakdown.values, 
                             textinfo='percent',
                             marker=dict(colors=['red', 'yellow', 'blue']))])

# Update layout
fig.update_layout(title='<b>Proportion of Total Costs in Food Delivery</b>', 
                  title_x=0.45,  # Title centered horizontally
                  showlegend=True)

# Show the plot
fig.show()

* Expense Allocation: A significant portion of the total costs is attributed to payment processing fees. This indicates the importance of considering payment processing costs in overall financial planning and management.

* Delivery Costs: The delivery fee amounts representing another substantial expense. Understanding and optimizing delivery costs can help streamline operations and improve cost efficiency.

* Discount Impact: The discount amount reflects the value of discounts applied, resulting in a reduction in total costs. While discounts can help attract customers and drive sales, it's essential to evaluate their impact on profitability and revenue generation.

* Cost Management: Analyzing cost breakdowns allows businesses to identify areas for cost optimization and efficiency improvements. 

* By monitoring and controlling expenses related to delivery and payment processing, businesses can enhance their financial performance and profitability.

#### Now, let's compare total revenue, total costs, and total profit (net loss in our case):

In [14]:
categories = ['Total Revenue', 'Total Costs', 'Total Profit']
values = [total_revenue, total_costs, total_profit]

# Create bar chart
fig = go.Figure(data=[go.Bar(x=categories, y=values, marker_color=['green', 'yellow', 'red'])])

# Update layout
fig.update_layout(title='<b>Comparison of Total Revenue, Total Costs, and Total Profit</b>',
                  xaxis_title='<b>Category</b>',
                  yaxis_title='Amount (INR)', title_x=0.45)
                 

# Show the plot
fig.show()

* Total costs surpass total revenue , leading to a substantial loss.
* This negative profit signals operational inefficiencies and poses risks to the financial health and sustainability of the business.

This situation underscores the importance of closely monitoring and managing costs to ensure they align with revenue generation. 

It also highlights the need for strategic decision-making to optimize operations, streamline expenses, and increase revenue streams. 

Without corrective actions, such as cost reduction initiatives or revenue enhancement strategies, the business may continue to experience financial losses, ultimately jeopardizing its long-term viability.

# Profit Enhancement Strategy:

#### To refine our profitability strategy, we've recognized the adverse impact of discounts on food orders, leading to substantial losses. 


Now, we aim to determine an optimal balance for offering discounts and charging commissions. 

#### This involves a deeper analysis of the traits of profitable orders. Specifically, we'll investigate:
* A new average commission percentage derived from profitable orders.
* A new average discount percentage for profitable orders, serving as a benchmark to gauge profitability.


#### With these fresh insights, we can propose adjustments that not only ensure profitability per order but also have broader applicability, enhancing overall profitability.
* The average commission percentage for profitable orders.
* The average discount percentage for profitable orders.

In [15]:
# Let's start by filtering the dataset for profitable orders
profitable_orders = data[data['Profit'] > 0].copy()

# calculate the average commission percentage for profitable orders
profitable_orders['Commission Percentage'] = (profitable_orders['Commission Fee'] / profitable_orders['Order Value']) * 100

# calculate the average discount percentage for profitable orders
profitable_orders['Effective Discount Percentage'] = (profitable_orders['Discount Amount'] / profitable_orders['Order Value']) * 100

# calculate the new averages
new_avg_commission_percentage = profitable_orders['Commission Percentage'].mean()
new_avg_discount_percentage = profitable_orders['Effective Discount Percentage'].mean()

new_avg_commission_percentage, new_avg_discount_percentage


(30.508436145149446, 5.867469879518072)

* #### New Average Commission Percentage: 30.51%
* #### New Average Discount Percentage: 5.87%

The average commission percentage for profitable orders is higher than the overall average for all orders. This suggests that charging a higher commission rate might be key to making orders profitable. 

On the other hand, the average discount percentage for profitable orders is lower than the overall average, indicating that offering lower discounts could still lead to profitability without losing too many orders.



### Based on this analysis, aiming for a commission rate around 31% and a discount rate around 6% could improve profitability overall.

Now, let's visualize how actual versus recommended discounts and commissions compare across all orders:

* #### First, calculate the profitability per order using the current discounts and commissions.
* #### Then, simulate profitability per order using recommended discounts (6%) and commissions (31%) to see how profitability might change.

This comparison will show the potential impact of using the recommended discount and commission rates on overall order profitability.

In [16]:
# simulate profitability with recommended discounts and commissions
recommended_commission_percentage = 31.0  # Recommended commission percentage (actual: 30.51%)
recommended_discount_percentage = 6.0    # Recommended discount percentage (actual: 5.86%)

# calculate the simulated commission fee and discount amount using recommended percentages
data['Simulated Commission Fee'] = data['Order Value'] * (recommended_commission_percentage / 100)
data['Simulated Discount Amount'] = data['Order Value'] * (recommended_discount_percentage / 100)

# recalculate total costs and profit with simulated values
data['Simulated Total Costs'] = (data['Delivery Fee'] + data['Payment Processing Fee'] + data['Simulated Discount Amount'])
data['Simulated Profit'] = (data['Simulated Commission Fee'] - data['Simulated Total Costs'])

# Create a subplot
fig = make_subplots(rows=1, cols=1)

# Actual profitability
fig.add_trace(go.Histogram(x=data['Profit'], histnorm='probability density', 
                            name='Actual Profitability', marker_color='skyblue',
                            opacity=0.5), row=1, col=1)

# Simulated profitability
fig.add_trace(go.Histogram(x=data['Simulated Profit'], histnorm='probability density', 
                            name='Estimated Profitability with Recommended Rates', marker_color='orange',
                            opacity=0.5), row=1, col=1)

# Update layout
fig.update_layout(title='<b>Actual vs. Recommended Discount and Commission</b>',
                  xaxis_title='Profit', yaxis_title='Density', title_x=0.45 )

# Show the plot
fig.show()


* The visualization compares the distribution of profitability per order using actual discounts and commissions versus the simulated scenario with recommended discounts (6%) and commissions (30%).

* The actual profitability distribution shows a mix, with a significant portion of orders resulting in losses (profit < 0) and a broad spread of profit levels for orders.

* The simulated scenario suggests a shift towards higher profitability per order. The distribution is more skewed towards positive profit, indicating that the recommended adjustments could lead to a higher proportion of profitable orders.

# CONCLUSION:


### Payment Method: 
* Cash on Delivery leads as the most preferred payment method, followed closely by Credit Card and Digital Wallet.

### Flexible Delivery Fees: 
* Customers are open to varying delivery fees, suggesting opportunities to adjust fee structures strategically to optimize revenue.

### Need for Continuous Improvement: 
* While most orders experience no refunds or chargebacks, ongoing enhancements in customer service processes are vital to minimize potential revenue losses.

### Profitability Challenges: 
* Despite a volume of orders (1,000), total costs exceed revenue, resulting in a net loss of -105,719.85 INR.

### Optimization Strategy: 
* To enhance profitability, businesses should aim for an optimal balance between discounts and commissions, with suggested average rates of 31% and 6%, respectively.